apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx

```diff
@@ -0,0 +1,308 @@
+      return {
+        selectedEnvironment: newSelectedEnvironment,
+        selectedCategory: newSelectedCategory,
+        categoriesAvailableForSelect: newCategoriesForSelect,
+        page: 1
+      };
+    }
+    case 'selectCategory':
+      return {
+        selectedEnvironment: state.selectedEnvironment,
+        selectedCategory: action.selection,
+        categoriesAvailableForSelect: action.categoriesAvailableForSelect,
+        page: 1
+      };
+    case 'selectPage':
+      return {
+        selectedEnvironment: state.selectedEnvironment,
+        selectedCategory: action.selectedCategory,
+        categoriesAvailableForSelect: action.categoriesAvailableForSelect,
+        page: action.selection
+      };
+    default:
+      throw new Error();
+  }
+}
+
+/*
+ * A component that fetches programming expressions and displays them in a
+ * table. It includes filters for which expressions are fetched as well as the
+ * ability to destroy and clone expressions.
+ * This is a levelbuilder-facing component.
+ */
+export default function ProgrammingExpressionsTable({
+  allProgrammingEnvironments,
+  allCategories,
+  hidden
+}) {
+  const [params, dispatchParams] = useReducer(paramsReducer, {
+    selectedEnvironment: DEFAULT_VALUE,
+    selectedCategory: DEFAULT_VALUE,
+    page: 1,
+    categoriesAvailableForSelect: allCategories
+  });
+
+  const [programmingExpressions, setProgrammingExpressions] = useState([]);
+  const [numPages, setNumPages] = useState(1);
+  const [error, setError] = useState(null);
+
+  const [itemToClone, setItemToClone] = useState(null);
+  const [itemToDelete, setItemToDelete] = useState(null);
+
+  const actionsCellFormatter = (actions, {rowData}) => {
+    return (
+      <div style={styles.actionsColumn}>
+        <Button
+          icon="edit"
+          text=""
+          href={rowData.editPath}
+          target="_blank"
+          __useDeprecatedTag
+          color="teal"
+        />
+        <Button
+          onClick={() => setItemToClone(rowData)}
+          text=""
+          icon="clone"
+          color="gray"
+          style={{margin: 0}}
+        />
+        <Button
+          onClick={() => setItemToDelete(rowData)}
+          text=""
+          icon="trash"
+          color="red"
+          style={{margin: 0}}
+        />
+      </div>
+    );
+  };
+
+  const fetchExpressions = (environmentId, categoryId, page) => {
+    const data = {};
+    if (environmentId !== DEFAULT_VALUE) {
+      data.programmingEnvironmentId = environmentId;
+    }
+    if (categoryId !== DEFAULT_VALUE) {
+      data.categoryId = categoryId;
+    }
+    data.page = page;
+    const url =
+      '/programming_expressions/get_filtered_expressions?' +
+      queryString.stringify(data);
+    let success = false;
+    fetch(url)
+      .then(response => {
+        if (response.ok) {
+          success = true;
+        } else {
+          setError(response.statusText);
+        }
+        return response.json();
```

Comment:
What happens if response.json() fails here (e.g. if the response body is not vaild json)?

I'm wondering if there are error cases in additional to the one handled on line 124 and if/how they should be handled.

Reply:
Good question. I'll add error handling.